features/steps/thens/values.py

```diff
@@ -1,86 +1,102 @@
 import csv
-import errors as err
 import ifcopenshell
 import os
 
-from behave import *
+from behave import register_type
 from pathlib import Path
-from utils import ifc, misc
 
+from validation_handling import gherkin_ifc, StepResult, handle_errors
 
-@then("The value must {constraint}")
-@then("The values must {constraint}")
-@then('At least "{num:d}" value must {constraint}')
-@then('At least "{num:d}" values must {constraint}')
-@err.handle_errors
-def step_impl(context, constraint, num=None):
-    errors = []
+from parse_type import TypeBuilder
+register_type(unique_or_identical=TypeBuilder.make_enum(dict(map(lambda x: (x, x), ("be unique", "be identical"))))) # todo @gh remove 'be' from enum values
+register_type(value_or_type=TypeBuilder.make_enum(dict(map(lambda x: (x, x), ("value", "type"))))) # todo @gh remove 'be' from enum values
+
+
+@gherkin_ifc.step("The value must be in '{csv_file}.csv'")
+@gherkin_ifc.step("The values must be in '{csv_file}.csv'")
+def step_impl(context, inst, csv_file):
+    if not inst:
+        return []
+
+    dirname = os.path.dirname(__file__)
+    filename = Path(dirname).parent.parent / "resources" / f"{csv_file}.csv"
+    valid_values = [row[0] for row in csv.reader(open(filename))]
+    invalid_values = [value for value in inst if value not in valid_values]
+    for value in invalid_values:
+        yield StepResult(expected = f"Value in {csv_file}.csv", observed = value)
+
+    return []
+
+
+@gherkin_ifc.step('At least "{num:d}" value must {constraint}')
+@gherkin_ifc.step('At least "{num:d}" values must {constraint}')
+def step_impl(context, inst, constraint, num):
+    stack_tree = list(
+        filter(None, list(map(lambda layer: layer.get('instances'), context._stack))))
+
+    if num is not None:
```

Comment:
`num` can never be None here

Reply:
I think this is a leftover from the previous big step_impl function in which we had multiple @given statements stacked on top of that step_impl function.. 
So in some cases, num could be None, simply because it was a different statement. 

Anyhow, it's removed 
[discard leftover previous big step_impl function; num is not None](https://github.com/buildingSMART/ifc-gherkin-rules/pull/119/commits/424adb816263cdd97fab7cc4044eba747a035ad4)